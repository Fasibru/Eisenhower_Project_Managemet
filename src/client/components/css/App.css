/* color schema matrix
#eec170 Middle Yellow Red
#f2a65a Sandy Brown
#f58549 Mandarin
#772f1a Liver
*/

/* color schema surface (greenish)
#CAD2C5 Pastel Gray
#84A98C Dark Sea Green
#52796f Hooker's Green
#354f52 Dark Slate Gray
#2f3e46 Japanese Indigo
*/

/* color schema
#648ca6 Rackley
#394263 Deep Koamaru
#8fd4d9 Middle Blue Green
(#011638 Maastricht Blue)
(#364156 Charcoal)
#EAE0CC Pearl
#C9ADA1 Silver Pink
#FBFCFF White
#F5FBFF Honeydew
#D0CCD0 Light Gray
*/

/* Global settings 
----------------------------------------------*/
body {
  font-family: 'Pontano Sans', sans-serif;
}
/* Grid settings
----------------------------------------------*/
.grid-container {
  display: grid;
  grid-template-columns: 200px 1fr;
  grid-template-rows: 150px 1fr 50px;
  grid-template-areas:
    "sidenav header"
    "sidenav main"
    "sidenav footer";
  height: 100vh;
}

/* Header 
----------------------------------------------*/
/* .header {
  grid-area: header;
  background-color: #394263;
} */

/* Sidenav 
----------------------------------------------*/
/* .sidenav {
  display: flex;
  flex-direction: column;
  grid-area: sidenav;
  background-color: #648ca6;
}

.sidenav__list {
  margin-top: 50px;
  padding-left: 20px;
  list-style-type: none;
}

.sidenav__list-elem {
  color: #f5fbef;
  padding: 20px 20px 20px 0px;
} */

/* Main 
----------------------------------------------*/
/* .main {
  grid-area: main;
  display: flex;
  background-color: #f5fbef;
}

.column-category {
  flex: 1;
  border-right: 1px solid black;
} */

/* Footer 
----------------------------------------------*/
/* .footer {
  grid-area: footer;
  background-color: #394263;
} */