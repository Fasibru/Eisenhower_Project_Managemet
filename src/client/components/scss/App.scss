/* color schema matrix
#eec170 Middle Yellow Red
#f2a65a Sandy Brown
#f58549 Mandarin
#772f1a Liver
*/

/* color schema surface (greenish)
#CAD2C5 Pastel Gray
#84A98C Dark Sea Green
#52796f Hooker's Green
#354f52 Dark Slate Gray
#2f3e46 Japanese Indigo
*/

/* color schema
#648ca6 Rackley
#394263 Deep Koamaru
#8fd4d9 Middle Blue Green
(#011638 Maastricht Blue)
(#364156 Charcoal)
#EAE0CC Pearl
#C9ADA1 Silver Pink
#FBFCFF White
#F5FBFF Honeydew
#D0CCD0 Light Gray
*/

/* Global settings 
----------------------------------------------*/
body {
  font-family: 'Pontano Sans', sans-serif;
}
/* Grid settings
----------------------------------------------*/
.grid-container {
  display: grid;
  grid-template-columns: 220px 1fr;
  grid-template-rows: 100px 1fr 50px;
  grid-template-areas:
    "sidenav header"
    "sidenav main"
    "sidenav footer";
  height: 100vh;
}