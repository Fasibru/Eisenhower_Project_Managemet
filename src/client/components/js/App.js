import React, { Component } from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { getTasks, toggleNewTaskPopup } from '../../actions/index';

import Header from './Header';
import Sidenav from './Sidenav';
import Main from './Main';
// import FilteredMain from '../../containers/FilteredMain';
import Footer from './Footer';
import NewTask from './NewTask';
import EditTask from './EditTask';

import '../css/App.css';

function mapStateToProps(state) {
  return {
    // filteredTasksRedux: state.filteredTasksRedux,
    // filtersRedux: state.filtersRedux,
    newTaskPopup: state.newTaskPopup,
  };
}

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      filteredTasks: [],
      newTask: {},
      editTaskPopup: false,
      editTask: {},
      filters: {},
    };
  }

  componentDidMount = () => {
    // read initial data form DB based on filters
    axios.get('/api/getInitialData')
      .then((res) => {
        this.setState({
          filters: res.data.filters[0],
          filteredTasks: res.data.tasks,
        });
      })
      .catch((err) => {
        console.log(err);
      });
    // const {
    //   // eslint-disable-next-line no-shadow
    //   // getTasks,
    // } = this.props;
    // getTasks();
  }

  // toggleNewTaskPopup = () => {
  //   const {
  //     // eslint-disable-next-line no-shadow
  //     toggleNewTaskPopup,
  //   } = this.props;
  //   toggleNewTaskPopup();
  // }

  toggleEditTaskPopup = () => {
    const { editTaskPopup } = this.state;
    // check if popup is already open and if yes then close it and reset editTask
    if (editTaskPopup) {
      this.setState({
        editTaskPopup: !editTaskPopup,
        editTask: {},
      });
    } else {
      this.setState({
        editTaskPopup: !editTaskPopup,
      });
    }
    // this.setState(prevState => ({ editTaskPopup: !prevState.editTaskPopup }));
  };

  populateEditTask = (data) => {
    // open popup
    this.toggleEditTaskPopup();
    // populate editTask based on data of double clicked task
    this.setState({
      editTask: {
        _id: data._id,
        rank: data.rank,
        category: data.category,
        title: data.title,
        description: data.description,
        completed: data.completed,
      },
    });
  }

  // add new task to data warehouse on submit
  handleNewTaskFormSubmit = (event) => {
    event.preventDefault();

    const { newTask, filteredTasks } = this.state;
    // eslint-disable-next-line no-shadow
    const { toggleNewTaskPopup } = this.props;
    const filteredTasksLength = filteredTasks.length;

    // POST new task and update state afterwards
    axios.post('/api/newTask', newTask)
      .then((res) => {
        // retrieve _id and rank from posted newTask
        const newTaskID = res.data._id;
        const newTaskRank = res.data.rank;
        const newTaskCompleted = res.data.completed;

        // Instead of using a GET for all tasks just add the new task to the filteredTasks array.
        // _id, rank and completed are updated manually.
        filteredTasks.push(newTask);
        filteredTasks[filteredTasksLength]._id = newTaskID;
        filteredTasks[filteredTasksLength].rank = newTaskRank;
        filteredTasks[filteredTasksLength].completed = newTaskCompleted;

        // update state including the new task for re-render
        this.setState({
          filteredTasks,
        });

        // reset newTask
        this.setState({
          newTask: {},
        });
      })
      .catch((err) => {
        console.log(err);
      });

    // close the popup after submit
    toggleNewTaskPopup();
  }

  // update newTask object based on form input in NewTask component:
  handleNewTaskFormChange = (event) => {
    const { newTask } = this.state;
    if (event.target.name === 'completed') {
      newTask[event.target.name] = event.target.checked;
    } else {
      newTask[event.target.name] = event.target.value;
    }

    // reflect form changes in newTask
    this.setState({
      newTask,
    });
  }

  handleEditTaskFormSubmit = (event) => {
    event.preventDefault();

    const { editTask, filteredTasks } = this.state;

    // save changes to the DB
    axios.put(`/api/editTask/${editTask._id}`, editTask)
      .catch((err) => {
        console.log(err);
      });

    // save changes in filteredTasks array as well to avoid additional GET of all tasks
    // find index of task to update:
    const editTaskIndex = filteredTasks.findIndex(task => task._id === editTask._id);
    // update filteredTasks array
    filteredTasks[editTaskIndex] = editTask;

    // reset editTask
    this.setState({
      editTask: {},
    });

    // close the popup after submit
    this.toggleEditTaskPopup();
  }

  // update editTask object based on form input in EditTask component:
  handleEditTaskFormChange = (event) => {
    const { editTask } = this.state;

    if (event.target.name === 'completed') {
      editTask[event.target.name] = event.target.checked;
    } else {
      editTask[event.target.name] = event.target.value;
    }

    // reflect form changes in editTask
    this.setState({
      editTask,
    });
  };

  // delete a task from within the EditTask popup
  handleDeleteTask = () => {
    const { editTask, filteredTasks } = this.state;

    axios.delete(`/api/deleteTask/${editTask._id}`)
      .catch((err) => {
        console.log(err);
      });

    // remove deleted task from filteredTasks array
    const deleteTaskIndex = filteredTasks.findIndex(task => task._id === editTask._id);
    filteredTasks.splice(deleteTaskIndex, 1);

    // reset editTask
    this.setState({
      editTask: {},
    });

    // close the popup after submit
    this.toggleEditTaskPopup();
  };

  // handle filter to show/hide completed tasks
  handleFilterShowTasks = (event) => {
    const { filters } = this.state;

    filters.showTasks = event.target.value;

    axios.put('/api/updateFilters', filters)
      .then((res) => {
        this.setState({
          filters: res.data.filters[0],
          filteredTasks: res.data.tasks,
        });
      })
      .catch((err) => {
        console.log(err);
      });
  }

  render() {
    const {
      filters,
      filteredTasks,
      // newTaskPopup,
      newTask,
      editTaskPopup,
      editTask,
    } = this.state;
    const {
      // filteredTasksRedux,
      // filtersRedux,
      // eslint-disable-next-line no-shadow
      toggleNewTaskPopup,
      newTaskPopup,
    } = this.props;
    const { title, description } = newTask;

    return (
      <div className="grid-container">
        <Header />
        <Sidenav
          toggleNewTaskPopup={toggleNewTaskPopup}
          filters={filters}
          handleFilterShowTasks={this.handleFilterShowTasks}
        />
        {/* <FilteredMain /> */}
        <Main
          filteredTasks={filteredTasks}
          populateEditTask={this.populateEditTask}
        />
        <Footer />
        {newTaskPopup
          && (
          <NewTask
            toggleNewTaskPopup={toggleNewTaskPopup}
            defineNewTask={this.handleNewTaskFormChange}
            submitNewTask={this.handleNewTaskFormSubmit}
            title={title}
            description={description}
          />
          )
        }
        {editTaskPopup
          && (
            <EditTask
              toggleEditTaskPopup={this.toggleEditTaskPopup}
              editTask={editTask}
              defineEditTask={this.handleEditTaskFormChange}
              submitEditTask={this.handleEditTaskFormSubmit}
              handleDeleteTask={this.handleDeleteTask}
            />
          )
        }
      </div>
    );
  }
}

App.propTypes = {
  // getTasks: PropTypes.func.isRequired,
  toggleNewTaskPopup: PropTypes.func.isRequired,
  newTaskPopup: PropTypes.bool.isRequired,
};

export default connect(mapStateToProps, { toggleNewTaskPopup })(App);
