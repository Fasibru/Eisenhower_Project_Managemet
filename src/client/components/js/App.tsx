// tslint:disable-next-line: import-name
import React, { useEffect, useState } from 'react';
// import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { ThunkAction } from 'redux-thunk';

import { getUserFilters } from '../../actions/actionsFilters';
import { getUserTasks } from '../../actions/actionsTasks';
import { getUser } from '../../actions/actionsUser';

import { FiltersActionsTypes } from '../../../types/filterActionTypes';
import { Filters, Store, TaskType, UserType } from '../../../types/storeTypes';
import { TaskActionsTypes } from '../../../types/taskActionTypes';
import { UserActionsTypes } from '../../../types/userActionTypes';

import MainContainer from '../../containers/Main.container';
import EditTask from './EditTask';
import Footer from './Footer';
import Header from './Header';
import LoadingScreen from './LoadingScreen';
import MenuBar from './MenuBar';
import NewTask from './NewTask';
import Sidenav from './Sidenav';

import '../scss/App.scss';

interface AppProps {
  newTaskPopup: boolean;
  editTaskPopup: boolean;
  userId: string;
  userError: string;
  isFetchingTasks: boolean;
  isFetchingUser: boolean;
  isFetchingFilters: boolean;
  getUserTasks(userId: string): ThunkAction<
    void,
    TaskType[],
    null,
    TaskActionsTypes
  >;
  getUserFilters(userId: string): ThunkAction<
    void,
    Filters,
    null,
    FiltersActionsTypes
  >;
  getUser(): ThunkAction<
    void,
    UserType,
    null,
    UserActionsTypes
  >;
}

const mapStateToProps = (state: Store) => ({
  editTaskPopup: state.tasks.editTaskPopup,
  isFetchingFilters: state.filters.isFetchingFilters,
  isFetchingTasks: state.tasks.isFetchingTasks,
  isFetchingUser: state.user.isFetchingUser,
  newTaskPopup: state.tasks.newTaskPopup,
  userError: state.user.userError,
  userId: state.user.userId,
});

// tslint:disable-next-line: variable-name
export const App: React.FC<AppProps> = ({
  // tslint:disable: no-shadowed-variable
  getUserTasks,
  getUserFilters,
  getUser,
  newTaskPopup,
  editTaskPopup,
  userId,
  userError,
  isFetchingTasks,
  isFetchingUser,
  isFetchingFilters,
  // tslint:enable: no-shadowed-variable
}) => {
  const [isInitialRender, setInitialRender] = useState(true);
  const [filtersMenuOpen, setFiltersMenuFlag] = useState(false);

  // guide on fetching data with useEffect:
  // https://www.robinwieruch.de/react-hooks-fetch-data/
  useEffect(() => {
    const fetchData = async () => {
      await getUser();
      if (userError) {
        history.pushState(null, null, '/');
        history.go();
      }
    };
    fetchData();
  },
  [userError]);

  useEffect(() => {
    if (userId) {
      getUserTasks(userId);
      getUserFilters(userId);
    }
    setInitialRender(false);
  },
  [userId]);

  if (isFetchingUser || isFetchingTasks || isFetchingFilters || isInitialRender) {
    return (
      <LoadingScreen />
    );
  }

  const toggleFilterMenu = () => {
    setFiltersMenuFlag(!filtersMenuOpen);
  };

  return (
    <div className="grid-container">
      <button
        type="button"
        id="sidenav__menu-icon"
        className="sidenav__menu-icon"
        // className={filtersMenuOpen ? 'sidenav__menu-icon' : 'sidenav__menu-icon sidenav--active'}
        onClick={toggleFilterMenu}
        onMouseDown={e => e.preventDefault()} /* to remove focus after button is clicked */
      >
        <MenuBar />
      </button>
      <Header />
      <Sidenav isOpen={filtersMenuOpen} toggleFilterMenu={toggleFilterMenu} />
      <MainContainer />
      <Footer />
      {newTaskPopup
        && (
          <NewTask />
        )
      }
      {editTaskPopup
        && (
          <EditTask />
        )
      }
    </div>
  );
};

// App.propTypes = {
//   getUserTasks: PropTypes.func.isRequired,
//   getUserFilters: PropTypes.func.isRequired,
//   getUser: PropTypes.func.isRequired,
//   newTaskPopup: PropTypes.bool.isRequired,
//   editTaskPopup: PropTypes.bool.isRequired,
//   userId: PropTypes.string.isRequired,
//   userError: PropTypes.string.isRequired,
//   isFetchingTasks: PropTypes.bool.isRequired,
//   isFetchingUser: PropTypes.bool.isRequired,
//   isFetchingFilters: PropTypes.bool.isRequired,
// };

export default connect(mapStateToProps, {
  getUser,
  getUserFilters,
  getUserTasks,
})(App);
