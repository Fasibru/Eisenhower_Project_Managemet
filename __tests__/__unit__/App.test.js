import React from 'react';
import { shallow } from 'enzyme';
import App from '../../src/client/components/js/App';

describe('Testing <App />', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  it('Renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
