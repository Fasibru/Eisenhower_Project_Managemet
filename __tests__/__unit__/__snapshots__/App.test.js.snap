// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <App /> Renders correctly 1`] = `
<div
  className="grid-container"
>
  <Header />
  <Sidenav
    filters={Object {}}
    handleFilterShowTasks={[Function]}
    toggleNewTaskPopup={[Function]}
  />
  <Main
    filteredTasks={Array []}
    populateEditTask={[Function]}
  />
  <Footer />
</div>
`;
